### DISCUSSION TITLE
Field offsets?

### DISCUSSION THREAD

**Discussion by untodesu** [dotnet/csharplang#6105] [Language Ideas] [answered]
C-family languages have a specific macro that allows you to get a byte offset for a certain field in a structure:
```c
struct funny_struct {
    int padding;
    int value;
};
size_t offset = offsetof(struct funny_struct, value); // 4
```

In terms of C#, Framework provides `Marshal.OffsetOf` method but it isn't reliable if we talk about managed structures.  
In any other case C# has no `offsetof` equivalent at the time.

What it could look like:
```csharp
struct FunnyStruct
{
    int Padding;
    int Value;
}

// ...
int offset = offsetof(FunnyStruct, Value); // 4
```

**Comment by jnm2** [comment]
How could `offsetof` be more reliable than `Marshal.OffsetOf`?

**Reply by untodesu** [reply]
We were discussing that with some people and it appears it may not be reliable in some cases (and if I understood correctly it `Marshal.OffsetOf` is an unmanaged offset which can be different):  

From docs.microsoft.com:
```
OffsetOf provides the offset in terms of the unmanaged structure layout,
which does not necessarily correspond to the offset of the managed structure layout.
```

If there's a way to use it reliably, I would really appreciate if you tell me about it...

**Reply by jnm2** [reply]
I'm pretty sure there's not a way, given that once the C# compiler is out of the picture, the runtime and platform determine the actual layout. Not an expert, but I know trivially how to create some cases where the offset is unknowable at compile time.

More specifically I'm curious what benefit `offsetof` brings versus using `Marshal.OffsetOf`. If you aren't using the structure in an unmanaged context, what is the use of knowing the offset?

**Reply by 333fred** [reply]
C# knows nothing about field layout or what offsets will be. That's all determined by the runtime, and `offsetof` wouldn't be able to do anything other than call `Marshall.OffsetOf`.

**Reply by untodesu** [reply]
So is `Marshal.OffsetOf` the only and best solution in terms of vertex formats then?

**Reply by jnm2** [reply]
Can you explain more about vertex formats? Are these structs going through some translation to the GPU or something? That would suggest perhaps a `SomeGpuLibrary.OffsetOf` method.

**Reply by untodesu** [reply]
I recently was experimenting with Silk.NET.OpenGL and there's not a single example in the repo where a vertices are represented as structures (instead, a simple float array is used).  

As you can guess, it's no good for a somewhat complex applications.  

OpenGL itself inputs vertex formats as a couple of values _including relative offset to each field_ that refers to the position in memory of a field if we are using a structure. And here's a thing: to achieve true modularity (say, using attributes on fields to generate vertex attributes using reflection) I need a way of getting this relative offset and relying on it all times.

**Reply by CyrusNajmabadi** [reply]
Why do you need a language feature here, instead of say a library feature that can answer this? 

This also feels like something an SG could provide, emitting the offsets as constants (or static readonly values) into some type you just reference by name.  E.g. Offsets.FunnyStruct_Value

**Reply by untodesu** [reply]
> Why do you need a language feature here, instead of say a library feature that can answer this?

If there's just a way to achieve this without a language feature then I guess don't need a language feature...

> This also feels like something an SG could provide, emitting the offsets as constants (or static readonly values) into some type you just reference by name. E.g. Offsets.FunnyStruct_Value

I really don't want to use source generators as they would complicate things

**Reply by jnm2** [reply]
It seems quite dangerous to be exposing managed memory in a raw form to OpenGL. If the structs are unmanaged (allowed by `where T : unmanaged`), then this is in unmanaged territory and Marshal.SizeOf seems right.

**Reply by untodesu** [reply]
Ah, fair enough then, thanks

**Comment by huoyaoyuan** [comment]
`Unsafe.ByteOffset(ref Unsafe.As<FunnyStruct, byte>(ref funnyStruct), ref Unsafe.As<int, byte>(ref funnyStruct.Value))`

JIT can't optimize this into a constant though.

**Reply by CyrusNajmabadi** [reply]
Why can't it?

**Reply by huoyaoyuan** [reply]
It doesn't optimize it now, but nothing prevents it.

**Reply by CyrusNajmabadi** [reply]
You said it "can't", so I thought there was an implication that there was something intrinsic about that that was blocking the runtime. Thanks for the clarification.

**Comment by untodesu** [comment, marked as answer]
jnm2 said:
> It seems quite dangerous to be exposing managed memory in a raw form to OpenGL. If the structs are unmanaged (allowed by `where T : unmanaged`), then this is in unmanaged territory and Marshal.SizeOf seems right.

aye fair enough

**Comment by The-Futurist** [comment]
The reservations expressed here refer to the danger of exposing details of managed memory but there's no reason to assume that this is going to refer to managed memory. We have the `unsafe` operator `sizeof` today that we can use for an `unmanaged` type so perhaps we can consider `offsetof` to tell us the offset of a member field within an `unmanaged` type.

Without this operator it is cumbersome at best to determine the offset of a member field and since the padding and alignment are potentially unpredictable it involves making some assumptions or performing pointer subtraction.

**Reply by CyrusNajmabadi** [reply]
@Korporal this thread already lists the way this can be done today for an unmanaged type.  To be a language feature an argument would have to be made as to why the existing API alarmed approaches are not suitable.

**Reply by The-Futurist** [reply]
Are you referring to this?

```cs
Unsafe.ByteOffset(ref Unsafe.As<FunnyStruct, byte>(ref funnyStruct), ref Unsafe.As<int, byte>(ref funnyStruct.Value))
```

**Reply by CyrusNajmabadi** [reply]
@Korporal Yes.  Though it's up to you how you want to expose/encapsulate things.  I have it abstracted so you can write:

```c#
GetFieldOffsetInBytes((ref FunnyStruct _) => ref _.Value)
``` 

But Marshal.OffsetOf works well too.

**Reply by The-Futurist** [reply]
Well the docs for `Marshal.OffsetOf` state that this gives an offset with respect to _marshalled structures_, not the layout of the type as managed by the CLR. 

> [OffsetOf](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.offsetof?view=net-8.0) provides the offset in terms of the unmanaged structure layout, which does not necessarily correspond to the offset of the managed structure layout.

Right there, that phrase differentiates between _two kinds_ of layout and hence two possible _different_ offset values.

We also have `Marshal.SizeOf` but despite that we still have the operator `sizeof`.  Does the `Marshal.OffsetOf` function give the same results as we get if we do a pointer subtraction in unsafe C# code? This is far from clear if we read the docs.

The suggestion for an `offsetof` operator is to give _exactly the same_ results as we'd get with C# unsafe pointer subtractions in a struct that satisfies the `unmanaged` constraint. If `Marshal.OffsetOf` does that for a struct that satisfies the `unmanaged` constraint then I'd be content with that, but the documentation for it says no such thing.

**Reply by CyrusNajmabadi** [reply]
Try out your scenarios and see :-)

Regardless, this feels firmly as an API question. I'd recommend conversing with the runtime team if you need more clarity or capabilities here. 

>  This is far from clear if we read the docs.

Doc improvement requests can go to the runtime, or the docs team.  Regardless, as above, this is well out of bounds of being a language question/discussion.

**Reply by The-Futurist** [reply]
> Try out your scenarios and see :-)

I'd rather have an answer from the C# team. Trying to infer what the system does is a recipe for disaster too, so that's pretty poor "advice".

> Regardless, this feels firmly as an API question. I'd recommend conversing with the runtime team if you need more clarity or capabilities here.

Will do.

> Doc improvement requests can go to the runtime, or the docs team. Regardless, as above, this is well out of bounds of being a language question/discussion.

Of course, but I don't know **IF** the documentation is at fault, I just know what it **doesn't say**, whether it should or should not is the reason I asked you. You clearly don't know the answer so I'll [reach out to people who will](https://github.com/dotnet/runtime/discussions/100487), given that you _don't know_ I wonder why you even participated in the discussion.

**Reply by CyrusNajmabadi** [reply]
> I'd rather have an answer from the C# team. 

The question is not a language one, but a runtime one.  As i mentioned, I'd recommend conversing with the runtime team if you need more clarity or capabilities here.

>  I wonder why you even participated in the discussion.

To direct you at the right people.  You're asking questions to the wrong team here.  The discussion is better suited to those who own this domain.

**Reply by CyrusNajmabadi** [reply]
> You clearly don't know the answer

I gave the answer on how the language thinks about it.  From an actual impl perspective, i also gave answers on potential options you can take, including the approach i use.  More details beyond that can go to the owners of those respective areas (like the doc owners or runtime owners).

**Reply by richlander** [reply]
@Korporal -- Your statements read like personal attacks. I'm certain I'm not the only one that feels that way. The guidance you were given was quite clearly given in good faith and (as a member of the team) seems like the correct advice. Please adopt a more professional style of communication. Your approach in this thread isn't helping anyone, including achieve your goals.

**Comment by The-Futurist** [comment]
@richlander 

> @Korporal -- Your statements read like personal attacks. I'm certain I'm not the only one that feels that way. The guidance you were given was quite clearly given in good faith and (as a member of the team) seems like the correct advice. Please adopt a more professional style of communication. Your approach in this thread isn't helping anyone, including achieve your goals.

With all due respect I stand by what I wrote, when a question about the specifics of some technology receives a dismissive response from a Microsoft representative of "Try out your scenarios and see :-)" I form the impression that my question is regarded as a joke, that's actually unprofessional, that's disparaging, that's elitism.

Read my first three posts in this thread and let me know if you think there was anything inappropriate there, anything that deserved a dismissive, disparaging response of "Try out your scenarios and see :-)" many of us that post here, start and participate in discussions are professionals and also customers; it sometimes seems to me that customer service isn't very well understood by some here.

Finally I was also told "But Marshal.OffsetOf works well too." which directly contradicts what other Microsoft representative say - [see here](https://github.com/dotnet/runtime/discussions/100487#discussioncomment-8975642), if you treat someone's questions as a joke and then mislead them they are unlikely to react with warmth and appreciation.

Tanner's replies are professional and objective, they do not disparage, they do not talk down to a person just because they have less expertise and knowledge.

**Comment by CyrusNajmabadi** [comment]
> disparaging response of "Try out your scenarios and see :-)"

This was a friendly suggestion, made in good faith.  Often it's good to just try things out and find out first hand what happens in the scenarios in question.  :)

> and then mislead them 

There was no misleading.  It's an option, with its own set of caveats (just like all things).  That's why i mentioned options to use, suggested you try them, but also pushed heavily that you seek out the best expertise on the topic (multiple times).  That's not misleading, that's sharing what is known but also recommending a great path to get the most knowledge possible to take on a challenge :)

Good luck! :)

**Reply by richlander** [reply]
In the interest of helping ... Here, this is directed at anyone that wants to listen.

If you see something you feel is disparaging and then (objectively) escalate the situation, then the entire burden of the problem conversation is on you. It's not a question of whether you accept this. This is "how it works".

Human interactions can be asymmetric (as is the case here). You often don't get an opportunity to reframe other people's perception of your behavior. The only good option is to play by the rules of the crowd. Your specific opinion on what is right and wrong doesn't matter or apply, unless you can get the crowd to listen. Again, this is "how it works", not just in this forum.
